starknet_py/net/full_node_client.py

```diff
@@ -0,0 +1,267 @@
+from typing import List, Optional, Union
+
+import aiohttp
+from marshmallow import EXCLUDE, ValidationError
+
+from starknet_py.net.base_client import (
+    BaseClient,
+    BlockHashIdentifier,
+    BlockNumberIdentifier,
+)
+from starknet_py.net.client_models import (
+    SentTransaction,
+    Transaction,
+    ContractCode,
+    TransactionReceipt,
+    BlockState,
+    StarknetBlock,
+    StarknetTransaction,
+    ContractDefinition,
+    InvokeFunction,
+    TransactionStatus,
+)
+from starknet_py.net.rpc_schemas.rpc_schemas import (
+    TransactionSchema,
+    TransactionReceiptSchema,
+    ContractCodeSchema,
+    StarknetBlockSchema,
+)
+from starknet_py.net.base_client_schemas import (
+    BlockHashIdentifierSchema,
+    BlockNumberIdentifierSchema,
+)
+from starknet_py.net.client_errors import ClientError
+
+
+class FullNodeClient(BaseClient):
+    def __init__(self, node_url):
+        self.url = node_url
+        self.rpc_client = RpcClient(url=node_url)
+
+    async def wait_for_tx(
+        self,
+        tx_hash: Union[int, str],
+        wait_for_accept: Optional[bool] = False,
+        check_interval=5,
+    ) -> (int, TransactionStatus):
+        pass
+
+    async def estimate_fee(self, tx: InvokeFunction) -> int:
+        pass
+
+    async def get_block(
+        self,
+        block_hash: Optional[Union[int, str]] = None,
+        block_number: Optional[int] = None,
+    ) -> StarknetBlock:
+        if block_hash is not None and block_number is not None:
+            raise ValueError(
+                "Block_hash and block_number parameters are mutually exclusive."
+            )
+
+        if block_hash is None and block_number is None:
+            raise ValueError("Block_hash or block_number must be provided.")
+
+        res = None
+        if block_hash is not None:
+            res = await self.rpc_client.call(
+                method_name="getBlockByHash",
+                params={
+                    "block_hash": str(hex(block_hash))
+                    if isinstance(block_hash, int)
+                    else block_hash,
+                    "requested_scope": "FULL_TXNS",
+                },
+            )
+        if block_number is not None:
+            res = await self.rpc_client.call(
+                method_name="getBlockByNumber",
+                params={"block_number": block_number, "requested_scope": "FULL_TXNS"},
+            )
+        return StarknetBlockSchema().load(res, unknown=EXCLUDE)
+
+    async def get_state_update(
+        self,
+        block_hash: Optional[Union[int, str]] = None,
+        block_number: Optional[int] = None,
+    ) -> BlockState:
+        # TODO when pathfinder node adds support (currently untestable)
+        pass
+
+    async def get_storage_at(
+        self,
+        contract_address: Union[int, str],
+        key: int,
+        block_hash: Optional[Union[int, str]] = None,
+        block_number: Optional[int] = None,
+    ) -> int:
+        if block_hash is None:
```

Comment:
block number is ignored in this function

Reply:
RPC doesn't support block number in this method

Reply:
Added en exception, but I'm unsure if handling it this way is the best for user experience. We could theoretically separate this into two methods and have only one in the `BaseSigner`. Which do you think is a better approach in that case?

Reply:
Hmm can you inherit a method in python and add additional parameters that would be recognised by type checker?

Reply:
From what I see it is possible to add addition parameters to the inherited method. I'm not sure if that's the ideal solution though?

Reply:
Fixed by simplifying interface and only adding additional parameters to GatewayClient.